### PR TITLE
Add try* variants of traits for controllers that support non-blocking…

### PR DESCRIPTION
**Pull request by lulf**
… operation

### CODE REVIEW

**Comment by lulf** [comment]
The goal is to implement these for nrf-sdc, so that one can take advantage of non-async send if available. 

Perhaps the read/write part of the controller should be split, as I don't have the use case for a non-blocking read.

**Comment by lulf** [comment]
@alexmoon I've found that I can live without the blocking commands, so I've changed this PR to only add try + blocking variants of the Controller trait.

File: src/controller.rs [hunk 1]
```diff
@@ -15,9 +15,13 @@ use crate::param::{RemainingBytes, Status};
 use crate::transport::Transport;
 use crate::{cmd, data, param, ControllerToHostPacket, FixedSizeValue, FromHciBytes, HostToControllerPacket};
 
-pub trait Controller {
+pub mod blocking;
+
+pub trait ErrorType {
     type Error: embedded_io::Error;
+}
```
Comment (alexmoon) [inline src/controller.rs:22]: Could we just use `embedded_io::ErrorType` here?

File: src/controller.rs [hunk 2]
```diff
@@ -33,6 +37,7 @@ pub enum CmdError<E> {
     Io(E),
 }
 
+/// An error type for try operations.
 impl<E> From<param::Error> for CmdError<E> {
```
Comment (alexmoon) [inline src/controller.rs:41]: I think this doc comment was meant for the `TryError` below?

**Review by alexmoon — COMMENTED** [review]
I think we also need to add a blocking version of the `Transport` trait, and change `ExternalController` to impl `Controller` when `T: Transport` and `blocking::Controller` when `T: blocking::Transport`.

**Comment by alexmoon** [comment]
> @alexmoon I've found that I can live without the blocking commands, so I've changed this PR to only add try + blocking variants of the Controller trait.

Great! That simplifies things considerably.

I suspect if we were to want to implement "blocking" commands, it should probably drop back to the normal HCI way of doing things where you do a blocking send of the command, but the `CommandComplete` or `CommandStatus` result comes back through a read at some other time. Otherwise you're requiring at least two threads of execution and a `Sync` implementation of the controller in order to receive and match up the command result to the blocking command execution.

**Pull request merged**